### PR TITLE
[ISSUE #1397]Add abnormal recover for dleger modle

### PR DESCRIPTION
**Pull request by duhenglucky**
## What is the purpose of the change

Add abnormal recover for dleger modle， [fix the issue 1397](https://github.com/apache/rocketmq/issues/1397)

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java [hunk 1]
```diff
@@ -261,24 +258,96 @@ private void recover(long maxPhyOffsetOfConsumeQueue) {
...
+                index = 0;
+                mmapFile = mmapFiles.get(index);
+            }
+
+            ByteBuffer byteBuffer = mmapFile.sliceByteBuffer();
+            long processOffset = mmapFile.getFileFromOffset();
+            long mappedFileOffset = 0;
+            while (true) {
+                DispatchRequest dispatchRequest = this.checkMessageAndReturnSize(byteBuffer, checkCRCOnRecover);
+                int size = dispatchRequest.getMsgSize();
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java]: This size does not contain dledger header, causing the process offset  to be wrong. Maybe should add DLedgerEntry.BODY_OFFSET.

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java [hunk 2]
```diff
@@ -261,24 +258,96 @@ private void recover(long maxPhyOffsetOfConsumeQueue) {
...
-                this.defaultMessageStore.truncateDirtyLogicFiles(maxPhyOffset);
+            if (maxPhyOffsetOfConsumeQueue >= processOffset) {
+                log.warn("maxPhyOffsetOfConsumeQueue({}) >= processOffset({}), truncate dirty logic files", maxPhyOffsetOfConsumeQueue, processOffset);
+                this.defaultMessageStore.truncateDirtyLogicFiles(processOffset);
             }
             return;
         }
         //Indicate that, it is the first time to load mixed commitlog, need to recover the old commitlog
         isInrecoveringOldCommitlog = true;
         //No need the abnormal recover
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java]: Comment and code are inconsistent.

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java [hunk 3]
```diff
@@ -246,13 +242,14 @@ public SelectMappedBufferResult getData(final long offset, final boolean returnF
...
             SelectMmapBufferResult sbr = mappedFile.selectMappedBuffer(pos);
-            return  convertSbr(truncate(sbr));
+            return convertSbr(truncate(sbr));
         }
 
         return null;
     }
 
-    private void recover(long maxPhyOffsetOfConsumeQueue) {
+    private void recoverDlegerAbnormally(long maxPhyOffsetOfConsumeQueue) {
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java]: Dledger misspelled.

**Comment by bxfjb** [comment]
I'm sry but is this patch merged?  
I think this is the solution for #7355

**Pull request closed without merging**